message-passing/safe-message-handlers/src/types.ts

```diff
@@ -28,3 +27,8 @@ export interface ClusterManagerWorkflowResult {
   numCurrentlyAssignedNodes: number;
   numBadNodes: number;
 }
+
+export enum ClusterState {
+  UP = 'UP',
```

Comment:
I think we needed a 3rd state here for "not started" otherwise start isn't idempotent.I think the state machine should be like

NOT STARTED -> STARTED -> CLOSED right now it is  STARTED <-> CLOSED

Reply:
> I think we needed a 3rd state here for "not started" otherwise start isn't idempotent

OK, what's the bad behavior you're seeing exactly? `startCluster` is idempotent currently, right -- after applying the operation once, from either initial state, you enter an infinite sequence of `UP` states:

<img width="300" alt="image" src="https://github.com/user-attachments/assets/c2467205-a19e-415d-a424-e1a4753027d9">

Reply:
In your code you can transition for `Up` to `Down` in `shutDownCluster`. That

Reply:
Right, so the state machine is like this; both operations are idempotent. I'm still not getting the problem:

<img width="401" alt="image" src="https://github.com/user-attachments/assets/6ae852ab-7ba8-4e75-af3a-25c8649d6ec6">

Reply:
you can run `startCluster` ,... other messages , and then `shutDownCluster` and `startCluster`. I think you will clear the state of the nodes

Reply:
Right, so the cluster completely forgets its state when it's shut down. It's always initialized when it starts up

```typescript
      this.state.clusterState = ClusterState.UP;
      for (let i = 0; i < 25; i++) {
        this.state.nodes.set(i.toString(), null);
      }
```

(the number of nodes 25 never changes). That's OK isn't it, for a toy example?

Reply:
Discussed offline. The issue was that I was using the math/CS definition of idempotent whereas in the context of Temporal activities the definition is more like "will never happen a second time having happened once". Agree it makes sense to make the operations idempotent in the Temporal Activityh sense; I've switched to a 3-state state space now. Thanks, interesting discussion!